Merge upstream prometheus/prometheus at `4ac7806`

gotjosh opened the pull request:
## Dependencies
- https://github.com/prometheus/prometheus/pull/13962
- https://github.com/prometheus/prometheus/pull/13961
- https://github.com/prometheus/prometheus/pull/13874
- https://github.com/prometheus/prometheus/pull/13873
- https://github.com/prometheus/prometheus/pull/13812
- https://github.com/prometheus/prometheus/pull/13811
- https://github.com/prometheus/prometheus/pull/13810
- https://github.com/prometheus/prometheus/pull/13867
- https://github.com/prometheus/prometheus/pull/13817
- https://github.com/prometheus/prometheus/pull/13952
- https://github.com/prometheus/prometheus/pull/13954

## Very relevant
- https://github.com/prometheus/prometheus/pull/13731 (AM change that needs a changelog entry)
- https://github.com/prometheus/prometheus/pull/13861 (this needs to be copied manually)
- https://github.com/prometheus/prometheus/pull/13974 (new metrics that do need a changelog entry) 

## Not relevant
- https://github.com/prometheus/prometheus/pull/13976 (bugfix in a command that is not used in mimir)
- https://github.com/prometheus/prometheus/pull/13958 (refactor)
- https://github.com/prometheus/prometheus/pull/13898 (docs)
- https://github.com/prometheus/prometheus/pull/13891 (docs)
- https://github.com/prometheus/prometheus/pull/13905 (CI changes)
- https://github.com/prometheus/prometheus/pull/13906 (CI changes)
- https://github.com/prometheus/prometheus/pull/13935 (unregistering TSDB metrics - I _think_ we don't have a concept of destroy so we might not need it?)
- https://github.com/prometheus/prometheus/pull/13919 (service discovery)
- https://github.com/prometheus/prometheus/pull/13923 (promQL parser basic sync)
- https://github.com/prometheus/prometheus/pull/13916

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/mimir-prometheus?pullRequest=614) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/mimir-prometheus?pullRequest=614) before we can accept your contribution.<br/>**5** out of **14** committers have signed the CLA.<br/><br/>:white_check_mark: suntala<br/>:white_check_mark: bboreham<br/>:white_check_mark: ywwg<br/>:white_check_mark: beorn7<br/>:white_check_mark: gotjosh<br/>:x: simonpasquier<br/>:x: mmorel-35<br/>:x: hanghuge<br/>:x: NeerajGartia21<br/>:x: GiedriusS<br/>:x: SuperQ<br/>:x: alanprot<br/>:x: wbh1<br/>:x: ArthurSens<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/mimir-prometheus?pullRequest=614) it.</sub>

gotjosh commented:
To open this PR I had to resolve a set of conflicts:

```git
Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)
        deleted by us:   .github/workflows/buf-lint.yml
        deleted by us:   .github/workflows/buf.yml
        deleted by us:   .github/workflows/ci.yml
        deleted by us:   .github/workflows/codeql-analysis.yml
        deleted by us:   .github/workflows/repo_sync.yml
        deleted by us:   .github/workflows/scorecards.yml
        both modified:   go.mod
        both modified:   go.sum
        both modified:   rules/manager_test.go
```

For the workflow files, I have removed them entirely and for the `go.mod` and `manager_test.go` I have attached the diff 

[diff-gotjosh-merge-upstream.txt](https://github.com/grafana/mimir-prometheus/files/15110694/diff-gotjosh-merge-upstream.txt)
